src/passes/reconstruct-stringify-walker.cpp

```diff
@@ -0,0 +1,193 @@
+  DBG(std::cout << "\n\nexistingBuilder: " << &existingBuilder
+                << " outlinedBuilder: " << &outlinedBuilder);
+}
+
+void ReconstructStringifyWalker::addUniqueSymbol(SeparatorReason reason) {
+  if (auto curr = reason.getFuncStart()) {
+    startExistingFunction(curr->func);
+    return;
+  }
+
+  // instrCounter is managed manually and incremented at the beginning of
+  // addUniqueSymbol() and visitExpression(), except for the case where we are
+  // starting a new function, as that resets the counters back to 0
+  instrCounter++;
+  if (state == InSkipSeq) {
+    return;
+  }
+
+  DBG(std::string desc);
+  IRBuilder* builder = state == InSeq ? &outlinedBuilder : &existingBuilder;
+  if (auto curr = reason.getBlockStart()) {
+    if (auto expr = builder->visitBlockStart(curr->block);
+        auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "Block Start for ";);
+  } else if (auto curr = reason.getIfStart()) {
+    if (auto expr = builder->visitIfStart(curr->iff);
+        auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "If Start for ";);
+  } else if (reason.getEnd()) {
+    if (auto expr = builder->visitEnd(); auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "End for ";);
+  } else {
+    DBG(desc = "addUniqueSymbol for unhandled instruction ";);
+  }
+  DBG(printAddUniqueSymbol(builder););
+}
+
+void ReconstructStringifyWalker::visitExpression(Expression* curr) {
+  maybeBeginSeq();
+
+  if (state == NotInSeq) {
+    if (auto val = existingBuilder.visit(curr); auto err = val.getErr()) {
+      Fatal() << err->msg;
+    }
+  }
+
+  if (state == InSeq) {
+    if (auto val = outlinedBuilder.visit(curr); auto err = val.getErr()) {
+      Fatal() << err->msg;
+    }
+  }
+
+  DBG(printVisitExpression(curr));
+  if (state == InSeq || state == InSkipSeq) {
+    maybeEndSeq();
+  }
+}
+
+void ReconstructStringifyWalker::startExistingFunction(Function* func) {
+  if (auto val = existingBuilder.build(); auto err = val.getErr()) {
+    Fatal() << err->msg;
+  }
+  if (auto val = existingBuilder.visitFunctionStart(func);
+      auto err = val.getErr()) {
+    Fatal() << err->msg;
+  }
+  instrCounter = 0;
+  seqCounter = 0;
+  state = NotInSeq;
+  DBG(std::cout << "\n\n$" << func->name << " Func Start " << &existingBuilder);
+}
+
+ReconstructStringifyWalker::ReconstructState
+ReconstructStringifyWalker::getCurrState() {
+  instrCounter++;
+  // We are either in a sequence or not in a sequence. If we are in a sequence
+  // and have already created the body of the outlined function that will be
+  // called, then we will skip instructions, otherwise we add the instructions
+  // to the outlined function. If we are not in a sequence, then the
+  // instructions are sent to the existing function.
+  if (seqCounter < sequences.size() &&
+      instrCounter >= sequences[seqCounter].startIdx &&
+      instrCounter <= sequences[seqCounter].endIdx) {
```

Comment:
Should this be a `<` rather than a `<=`? I would have expected `endIdx` to be one-past-the-end so that `endIdx - startIdx = length`.

Reply:
you are quite right.